torch/csrc/distributed/rpc/message.cpp

```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
 
 std::vector<c10::weak_intrusive_ptr<c10::StorageImpl>> Message::getStorages()
     const {
+  // Sparse tensors do not have storage. Instead, a sparse tensor
+  // contains two tensors indices and values, and both contain storage.
   std::vector<c10::weak_intrusive_ptr<c10::StorageImpl>> storages;
-  storages.reserve(tensors_.size());
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  auto numStorages = tensors_.size();
+  for (const at::Tensor& tensor : tensors_) {
+    if (tensor.is_sparse()) {
+      ++numStorages;
+    }
+  }
+  storages.reserve(numStorages);
```

Comment:
Just curious, does extra loop over tensors outweighs perf benefits of doing something like `storages.reserve(tensors_.size());` or `storages.reserver(2*tensors_.size());`?

Reply:
I did a basic execution time experiment using chrono::high_resolution_clock. r2x measures the time it takes to reserve 2 * x elements. looprx measures the time it takes to loop x times and checks if a tensor is sparse. In the test, the tensor is not sparse.

```
1-r2x 227nanoseconds
1-looprx 96nanoseconds
10-r2x 135nanoseconds
10-looprx 73nanoseconds
100-r2x 2100nanoseconds
100-looprx 82nanoseconds
1000-r2x 2204nanoseconds
1000-looprx 1888nanoseconds
10000-r2x 4171nanoseconds
10000-looprx 3093nanoseconds
100000-r2x 4243nanoseconds
100000-looprx 5579nanoseconds
1000000-r2x 6117nanoseconds
1000000-looprx 4935nanoseconds
10000000-r2x 6836nanoseconds
10000000-looprx 4188nanoseconds
```

Worst case, the extra loop should be slower since it reserves the same amount. 

Below is measurements for a loop and check if a tensor is sparse.

```
1-loopx 21nanoseconds
10-loopx 21nanoseconds
100-loopx 21nanoseconds
1000-loopx 22nanoseconds
10000-loopx 22nanoseconds
100000-loopx 22nanoseconds
1000000-loopx 21nanoseconds
10000000-loopx 21nanoseconds
```

The loop doesn't cost much in terms of time. 

Also, I think dense tensors are used more than sparse tensors, so based on the basic experiments, I would not reserve 2x.

Reply:
I did a basic execution time experiment using chrono::high_resolution_clock. r2x measures the time it takes to reserve 2 * x elements. looprx measures the time it takes to loop x times and checks if a tensor is sparse. In the test, the tensor is sparse:

```
1-loopx 28 nanoseconds
1-r2x 198 nanoseconds
10-loopx 21 nanoseconds
10-r2x 2514 nanoseconds
100-loopx 21 nanoseconds
100-r2x 77 nanoseconds
1000-loopx 21 nanoseconds
1000-r2x 1951 nanoseconds
10000-loopx 22 nanoseconds
10000-r2x 3536 nanoseconds
100000-loopx 22 nanoseconds
100000-r2x 3962 nanoseconds
1000000-loopx 21 nanoseconds
1000000-r2x 5841 nanoseconds
10000000-loopx 21 nanoseconds
10000000-r2x 5119 nanoseconds
```

Worst case, the extra loop should be slower since it reserves the same amount. 

The loop doesn't cost much in terms of time. 

Also, I think dense tensors are used more than sparse tensors, so based on the basic experiments, I would not reserve 2x.

Reply:
I think there is an error in the loopx timing.

Reply:
```
  if (count > 0){
        for (int i = 1; i < 10000000; i*=10){
          const auto& begin = std::chrono::high_resolution_clock::now();
          int ns = 1;
          for(int j = 0; j < i; j++){
            if (save.is_sparse()){
              ns++;
            }
          }
          const auto& end = std::chrono::high_resolution_clock::now();
          std::cout << i << "-loopx " << std::chrono::duration_cast<std::chrono::nanoseconds>(end - begin).count() << "nanoseconds" << std::endl;
        }
  }
 ```
 
 Idk I guess its right actually 
 ```
1-loopx 64nanoseconds
10-loopx 33nanoseconds
100-loopx 32nanoseconds
1000-loopx 33nanoseconds
10000-loopx 32nanoseconds
100000-loopx 32nanoseconds
1000000-loopx 32nanoseconds
1-loopx 41nanoseconds
10-loopx 33nanoseconds
100-loopx 33nanoseconds
1000-loopx 32nanoseconds
10000-loopx 33nanoseconds
100000-loopx 32nanoseconds
1000000-loopx 32nanoseconds
1-loopx 72nanoseconds
10-loopx 32nanoseconds
100-loopx 33nanoseconds
1000-loopx 32nanoseconds
10000-loopx 32nanoseconds
100000-loopx 33nanoseconds
1000000-loopx 32nanoseconds

```